File: sources/CSharp/src/Ers/Licensing/License.cs
using Ers.Engine;

namespace Ers
{
    public static class License
    {
        public static bool HasFeature(string featureCode)
        {
            var featureCodeUtf8 = featureCode.ToUtf8NullTerminated();
            unsafe
            {
                fixed(byte* featureCodeByte = featureCodeUtf8)
                {
                    return ErsEngine.ERS_License_HasFeature(featureCodeByte);
                }
            }
        }

        public static string EditionName
        {
            get {
                unsafe
                {
                    char* heapAllocatedName = (char*)ErsEngine.ERS_License_EditionName();
                    string edition          = new string(heapAllocatedName);
                    ErsEngine.ERS_String_Destroy((nint)heapAllocatedName);
                    return edition;
                }
            }
        }

        public static int MaxJobSystemCores() { return ErsEngine.ERS_License_GetLimits_MaxJobSystemCores(); }
        public static int MaxComponentTypes() { return ErsEngine.ERS_License_GetLimits_MaxComponentTypes(); }
    }
}

File: sources/CPP/include/Ers/Licensing/License.cpp
#include "License.h"
#include "Ers/Api.h"

bool Ers::License::HasFeature(const char* featureCode)
{
    return Ers::Engine::ERS_License_HasFeature(featureCode);
}

std::string Ers::License::EditionName()
{
    char* heapAllocatedName = Ers::Engine::ERS_License_EditionName();
    std::string edition(heapAllocatedName);
    Ers::Engine::ERS_String_Destroy(heapAllocatedName);

    return edition;
}

int Ers::License::MaxJobSystemCores()
{
    return Ers::Engine::ERS_License_GetLimits_MaxJobSystemCores();
}

int Ers::License::MaxComponentTypes()
{
    return Ers::Engine::ERS_License_GetLimits_MaxComponentTypes();
}
